### PR TITLE
Backstop emissions refactor

### LINKED ISSUE #8: Reward Zone Limit
The reward zone limit is poorly designed, and will eventually outpace limit growth, causing a scenario that could be reached where the backstop is no longer able to emit to all reward zone pools.

There should be a hard cap on reward zone pools, secondly, we should look to reduce the ledger entry impact during `distribute` before this limit is computed.

Reward zone pools should also be able to be removed if they have less than 5% of the backstop threshold.

### PR DESCRIPTION
**Pull request by Ryang-21**
- Separate backstops emission distribution from emission gulping to prevent exceeding resource limits
- Increase emission decimals to 14 to reduce rounding loss
- Implement remove_reward to remove backstops that are below the threshold from the reward zone
fixes #8

### CODE REVIEW

File: backstop/src/emissions/manager.rs [hunk 1]
```diff
@@ -164,21 +209,29 @@ pub fn set_backstop_emission_eps(
...
+                .unwrap_optimized();
+            println!("tokens since last emission: {}", tokens_since_last_emission);
             tokens_left_to_emit += tokens_since_last_emission;
         }
-        let eps = u64(tokens_left_to_emit / (7 * 24 * 60 * 60)).unwrap_optimized();
+        // Scale eps by 14 decimal places to reduce rounding errors
+        let eps = u64(tokens_left_to_emit * SCALAR_7 / (7 * 24 * 60 * 60)).unwrap_optimized();
         emission_data.eps = eps;
         emission_data.expiration = expiration;
+        println!("new eps: {}\n new tokens: {}", eps, tokens_left_to_emit);
```
Comment (mootz12) [inline backstop/src/emissions/manager.rs:228]: println

File: backstop/src/emissions/distributor.rs [hunk 2]
```diff
@@ -70,15 +72,22 @@ pub fn update_emission_data(
...
             let new_data = BackstopEmissionData {
                 eps: emis_data.eps,
                 expiration: emis_data.expiration,
                 index: additional_idx + emis_data.index,
                 last_time: e.ledger().timestamp(),
             };
+            println!(
+                "expiration: {} , index: {}, additional_index: {}, unqueued_shares: {}",
+                new_data.expiration, new_data.index, additional_idx, unqueued_shares
+            );
```
Comment (mootz12) [inline backstop/src/emissions/distributor.rs:90]: println

File: backstop/src/emissions/manager.rs [hunk 3]
```diff
@@ -39,7 +41,69 @@ pub fn add_to_reward_zone(e: &Env, to_add: Address, to_remove: Address) {
...
+pub fn remove_from_reward_zone(e: &Env, to_remove: Address) {
+    let mut reward_zone = storage::get_reward_zone(e);
+
+    // enusre to_add has met the minimum backstop deposit threshold
+    // NOTE: "to_add" can only carry a pool balance if it is a deployed pool from the factory
+    let pool_data = load_pool_backstop_data(e, &to_remove);
+    if require_pool_above_threshold(&pool_data) {
+        panic_with_error!(e, BackstopError::BadRequest);
+    } else {
         let to_remove_index = reward_zone.first_index_of(to_remove.clone());
```
Comment (mootz12) [inline backstop/src/emissions/manager.rs:107]: can the core logic of `remove_from_reward_zone` be shared?

IE

```rust
/// Remove pool from the reward zone vec, does not store
fn remove_pool(reward_zone: &mut Vec<Address>, to_remove: &Address) {
    // check for existance of pool in the rz
    // check last distro time was recent
   // update rx_emissions
  // remove from rz_vec
}
```

Then both `add_to_rz` and `remove_from_rz` can share the most important chunk of logic when removing a pool.

File: backstop/src/emissions/manager.rs [hunk 4]
```diff
@@ -1,26 +1,28 @@
...
+pub fn add_to_reward_zone(e: &Env, to_add: Address, to_remove: Option<Address>) {
     let mut reward_zone = storage::get_reward_zone(e);
+    let rz_emission_index = storage::get_rz_emission_index(e);
     let max_rz_len = if e.ledger().timestamp() < BACKSTOP_EPOCH {
         10
     } else {
-        10 + (i128(e.ledger().timestamp() - BACKSTOP_EPOCH) >> 23) // bit-shift 23 is ~97 day interval
+        // Max reward zone length is 50
+        (10 + (i128(e.ledger().timestamp() - BACKSTOP_EPOCH) >> 23)).min(50) // bit-shift 23 is ~97 day interval
     };
```
Comment (mootz12) [inline backstop/src/emissions/manager.rs:26]: this should be fixed to 50, we can also get rid of the `BACKSTOP_EPOCH` variable

File: backstop/src/emissions/manager.rs [hunk 5]
```diff
@@ -83,70 +146,86 @@ pub fn gulp_emissions(e: &Env) -> i128 {
...
+                    .fixed_mul_floor(gulp_index - emission_data.index, SCALAR_14)
+                    .unwrap_optimized();
+                accrued += new_emissions;
+                return set_rz_emissions(e, pool, gulp_index, accrued, to_gulp);
+            } else {
+                return set_rz_emissions(e, pool, gulp_index, accrued, to_gulp);
+            }
+        }
     }
+    return 0;
```
Comment (mootz12) [inline backstop/src/emissions/manager.rs:218]: this should panic instead of return zero.

- [ ] Either way, can we double check that `gulp_emissions` on the pool can handle the case where it gets 0 tokens to emit?

File: pool/src/emissions/distributor.rs [hunk 6]
```diff
@@ -399,6 +401,66 @@ mod tests {
...
+            min_persistent_entry_ttl: 10,
+            max_entry_ttl: 3110400,
+        });
+        // Supply of 1 trillion at 18 decimals
+        let supply: i128 = 1000000000000_000000000000000000;
+        let user_position: i128 = 1_000000000000000000;
+        e.as_contract(&pool, || {
+            let reserve_emission_data = ReserveEmissionData {
+                expiration: 1600000000,
+                eps: 0_01000000000000,
```
Comment (mootz12) [inline pool/src/emissions/distributor.rs:428]: can this test emit significantly more tokens?

the overflow will occur on `emissions * supply_scalar * SCALAR_7`. The this looks like it emits 10,000 BLND, which means the math will fit in i128 (intermediate math is ~1e36 i think).

If you emit like 1m BLND it should break the i128 math, and force it to scale to i256.

File: pool/src/pool/user.rs [hunk 7]
```diff
@@ -709,17 +709,17 @@ mod tests {
             assert_eq!(reserve_0.b_supply, starting_b_token_supply + 123);
 
             let new_emis_res_data = storage::get_res_emis_data(&e, &res_0_d_token_index).unwrap();
-            let new_index = 1000
+            let new_index = 10000000000
                 + (1000i128 * 0_1000000)
-                    .fixed_div_floor(starting_b_token_supply, SCALAR_7)
+                    .fixed_div_floor(starting_b_token_supply, SCALAR_7 * SCALAR_7)
```
Comment (mootz12) [inline pool/src/pool/user.rs:714]: SCALAR_14

File: test-suites/tests/test_backstop_rz_changes.rs [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+
+    fixture
+        .backstop
+        .add_reward(&pool_fixture.pool.address, &None);
+
+    fixture.emitter.distribute();
+    fixture.backstop.distribute();
+
+    let result = pool_fixture.pool.gulp_emissions();
+    assert_eq!(result, 1814402999999); // (60 * 60 * 24 * 7 + 1) * 0.3
```
Comment (mootz12) [inline test-suites/tests/test_backstop_rz_changes.rs:96]: this assert is a bit misleading, given pools should not earn emissions when removed from the reward zone, but this is using an edge case in the emissions logic where if no eligible pools are in the reward zone, it will hold all incoming emissions until a pool is added.

At the very least, this is due a comment as the result is confusing.

File: test-suites/tests/test_wasm_happy_path.rs [hunk 9]
```diff
@@ -433,8 +433,8 @@ fn test_wasm_happy_path() {
         &vec![&fixture.env, pool_fixture.pool.address.clone()],
         &frodo,
     );
-    assert_eq!(claim_amount, 22014719_9950114); //actual amount is 22014720_0000000 but get's rounded down
-    backstop_blnd_balance -= 22014719_9950114;
+    assert_eq!(claim_amount, 22014719_9999997); //actual amount is 22014720_0000000 but get's rounded down
+    backstop_blnd_balance -= 22014719_9999997;
```
Comment (mootz12) [inline test-suites/tests/test_wasm_happy_path.rs:437]: much closer!

**Review by mootz12 — CHANGES_REQUESTED** [review]

File: backstop/src/emissions/manager.rs (hunk 5 above)
Comment (mootz12) [reply backstop/src/emissions/manager.rs:218]: looks like `gulp_emissions` will panic if less than 1 BLND is emitted, so this looks OK.

File: pool/src/pool/user.rs (hunk 7 above)
Comment (Ryang-21) [reply pool/src/pool/user.rs:714]: This would be the only place on the pool that would use SCALAR_14 so it seemed unnecessary to add it for just a test.

**Review by mootz12 — APPROVED** [review]
LG2M
